index.js

```diff
@@ -7,27 +7,73 @@
 
 'use strict';
 
-module.exports = function(prefix) {
-  var debug = require('debug');
-
-  return function(app) {
-    this.define('debug', function(name) {
-      var fn = debug(namespace(prefix, name));
-      return function() {
-        return fn.apply(debug, arguments);
-      };
+/**
+ * Plugin that adds `debug` method to the instance
+ * and `namespaces` are added to this method.
+ *
+ * ```js
+ * app.debug()('debugging on main namespace')
+ * //=> base debugging on main namespace
+ *
+ * app.debug.one = app.debug('one')
+ * app.debug.one('testing %s, whatever', 'foo, bar')
+ * //=> base:one testing foo, bar, whatever
+ *
+ * app.debug.mix = app.debug('two', 123 'three:four', 'five')
+ * app.debug.mix('okey, %s awesome', 'this is')
+ * //=> base:two:three:four:five okey, this is awesome
+ *
+ * app.debug.helper('is one of internal namespaces added')
+ * //=> base:helper is one of internal namespaces added
+ * ```
+ *
+ * @name   debug
+ * @param  {Array} `namespaces`
+ * @return {Function}
+ */
+
+module.exports = function debugPlugin(namespaces) {
+  namespaces = Array.isArray(namespaces) ? namespaces : [namespaces];
+
+  return function plugin(app) {
+    app.define('_debugPrefix', app._namespace);
+    app.define('debug', function debug() {
+      return debugFactory.apply(app, arguments);
     });
+    if (namespaces.length) {
+      var len = namespaces.length;
+      var i = 0;
+
+      while (i < len) {
+        var ns = namespaces[i++];
+        app.debug[ns] = debugFactory.call(app, ns);
+        app.debug[ns].color = app.debug.color;
+      }
+    }
```

Comment:
_(this is for the `templates` docs, but it will help explain how plugins work in `base`. you obviously already know a lot about plugins, so this is for anyone who might find it useful)_

# Plugins

> Overview of how `Templates` plugins work

## App

The following example shows a `plugin` that will be invoked by the `.use` method. Example: `app.use(plugin)`. When invoked, the plugin function is passed the application instance (`app`) as its only argument.

![plugins-app](https://cloud.githubusercontent.com/assets/383994/13402852/311b9d88-dee0-11e5-944b-200ba56f42fe.png)

The plugin stops there and will not be called again, **unless the plugin returns a function**. If a function is returned it will be pushed onto the `app.fns` array and then called on each collection that is created using the `app.create`, `app.collection`, or `app.list` methods.

In the next example, a function is returned from the plugin so that it can be called again later.

## Collection

![plugins-collection](https://cloud.githubusercontent.com/assets/383994/13402856/34dd996c-dee0-11e5-8def-f0b739ff10ca.png)

The plugin stops there and will not be called again, **unless the plugin returns a function**. If a function is returned it will be pushed onto the `collection.fns` array and then called on each view that is added to the collection.

In the next example, we'll return a function inside the `collection` plugin, so that it can be called again later.

## View

![plugins-view](https://cloud.githubusercontent.com/assets/383994/13402861/37fc7032-dee0-11e5-8489-04f392cb9370.png)

End of the line...

# Short-circuit: "smart plugins"

> Don't like all the function nesting? Want to register your plugin with `app` but only have it run on specific objects? No problem, just short-circuit the plugin!

Every class has a boolean `.is*` property that is useful for checking the instance name. For example, the `Templates` class has an `isTemplates` property, view collections have `isViews`, and views have `isView`. (all "apps", like `Templates`, `Assemble`, `Generate`, etc. also have `isApp` as a convenience).

To make your plugins "smarter" and eliminate the nesting, try the following:

![plugins-short-circuit](https://cloud.githubusercontent.com/assets/383994/13405180/33cea2a4-deeb-11e5-864d-4fbab7510c22.png)

TBC...